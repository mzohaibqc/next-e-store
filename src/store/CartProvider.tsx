'use client';

import React, {
  Dispatch,
  ProviderProps,
  createContext,
  useCallback,
  useReducer,
} from 'react';

import type { Product } from '@/components/ProductItem';

type Resolve<T> = {
  [K in keyof T]: Resolve<T[K]>;
};

export type CartItem = Resolve<Product & { count: number }>;

type CartAction = {
  type: 'ADD_ITEM' | 'REMOVE_ITEM' | 'CLEAR_ITEM';
  payload: Product;
};

export type CartStoreState = {
  total: number;
  cart: CartItem[];
  color?: string;
};
export type CartStoreType = CartStoreState & {
  addItem: (item: Product) => void;
  removeItem: (item: Product) => void;
  clearItem: () => void;
};
export const cartInitialState: CartStoreState = {
  total: 0,
  cart: [],
};

function updateCartItemCount(
  state: CartStoreState,
  id: number,
  count: number
): CartStoreState['cart'] {
  return state.cart.map((cartItem) => {
    if (cartItem.id === id) {
      return {
        ...cartItem,
        count: cartItem.count + count,
      };
    }
    return cartItem;
  });
}
function cartReducer(
  state: CartStoreState,
  action: CartAction
): CartStoreState {
  const item = action.payload;
  const existingItem = state.cart.find((cartItem) => cartItem.id === item.id);
  switch (action.type) {
    case 'ADD_ITEM':
      if (existingItem) {
        return {
          ...state,
          total: state.total + item.price,
          cart: updateCartItemCount(state, item.id, 1),
        };
      }
      return {
        ...state,
        total: state.total + item.price,
        cart: [...state.cart, { ...item, count: 1 }],
      };
    case 'REMOVE_ITEM':
      if (!existingItem) {
        return state;
      }
      return {
        ...state,
        total: state.total - existingItem.price,
        cart: updateCartItemCount(state, existingItem.id, -1),
      };
    case 'CLEAR_ITEM':
      if (!existingItem) {
        return state;
      }
      return {
        ...state,
        total: state.total - existingItem.count * existingItem.price,
        cart: state.cart.filter((cartItem) => cartItem.id !== existingItem.id),
      };

    default:
      return state;
  }
}

function useCartContext() {
  const [state, dispatch]: [CartStoreState, Dispatch<CartAction>] = useReducer(
    cartReducer,
    cartInitialState
  );
  const addItem = useCallback(
    (payload: Product) => dispatch({ type: 'ADD_ITEM', payload }),
    []
  );
  const removeItem = useCallback(
    (payload: Product) => dispatch({ type: 'REMOVE_ITEM', payload }),
    []
  );
  const clearItem = useCallback(
    (payload: Product) => dispatch({ type: 'CLEAR_ITEM', payload }),
    []
  );
  return { ...state, addItem, removeItem, clearItem };
}

export const CartContext = createContext<ReturnType<typeof useCartContext>>({
  total: 0,
  cart: [],
  addItem: () => {},
  removeItem: () => {},
  clearItem: () => {},
});

type Props = Omit<ProviderProps<CartStoreState>, 'value'>;

export function CartProvider({ children }: Props) {
  const state = useCartContext();
  return <CartContext.Provider value={state}>{children}</CartContext.Provider>;
}

export function useCart() {
  const store = React.useContext(CartContext);
  return store;
}
