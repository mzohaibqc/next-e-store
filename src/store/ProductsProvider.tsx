'use client';

import React, {
  Dispatch,
  ProviderProps,
  createContext,
  useCallback,
  useReducer,
} from 'react';
import type { Product } from '@/components/ProductItem';

export type ProductFilterAction = {
  type: 'SET_COLOR';
  payload: string | undefined;
};

export type SetProductsAction = {
  type: 'SET_PRODUCTS';
  payload: Product[];
};

export type ProductsAction = SetProductsAction | ProductFilterAction;

export type ProductStoreState = {
  products: Product[];
  color: string | undefined;
  filteredProducts: Product[];
};

export const initialProductsState: ProductStoreState = {
  products: [],
  color: undefined,
  filteredProducts: [],
};
export function productsReducer(
  state: ProductStoreState,
  action: ProductsAction
): ProductStoreState {
  switch (action.type) {
    case 'SET_PRODUCTS':
      return {
        ...state,
        products: action.payload,
        filteredProducts: action.payload,
      };
    case 'SET_COLOR':
      return {
        ...state,
        color: action.payload,
        filteredProducts: state.products.filter(
          (product) =>
            action.payload === undefined || product.colour === action.payload
        ),
      };

    default:
      return state;
  }
}

function useProductsContext({ products }: { products: Product[] }) {
  const [state, dispatch]: [ProductStoreState, Dispatch<ProductsAction>] =
    useReducer(productsReducer, {
      ...initialProductsState,
      products,
      filteredProducts: products,
    });
  const setProducts = useCallback(
    (payload: Product[]) => dispatch({ type: 'SET_PRODUCTS', payload }),
    []
  );
  const setColor = useCallback(
    (payload: string | undefined) => dispatch({ type: 'SET_COLOR', payload }),
    []
  );
  return { ...state, setProducts, setColor };
}

export const ProductsContext = createContext<
  ReturnType<typeof useProductsContext>
>({
  ...initialProductsState,
  setProducts: () => {},
  setColor: () => {},
});

type Props = Omit<ProviderProps<ProductStoreState>, 'value'> & {
  products: Product[];
};
export function ProductsProvider({ children, products }: Props) {
  const state = useProductsContext({ products });
  return (
    <ProductsContext.Provider value={{ ...state, products }}>
      {children}
    </ProductsContext.Provider>
  );
}

export function useProductsStore() {
  const store = React.useContext(ProductsContext);
  return store;
}
