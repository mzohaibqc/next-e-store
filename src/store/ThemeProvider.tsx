'use client';
import { ThemeProvider } from 'next-themes';
import { ThemeProviderProps } from 'next-themes/dist/types';

export function Provider(props: ThemeProviderProps) {
  return (
    <ThemeProvider {...props} attribute="class">
      {props.children}
    </ThemeProvider>
  );
}
