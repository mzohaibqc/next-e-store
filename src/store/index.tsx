'use client';
export * from './CartProvider';
export * from './ProductsProvider';
