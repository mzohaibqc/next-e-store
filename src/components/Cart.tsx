'use client';

import { useCart } from '@/store';

export default function Cart() {
  const { total } = useCart();
  return (
    <div className="w-full lg:w-[300px] lg:h-[400px] p-4 rounded-md border bg-white flex justify-between">
      <p className="text-lg font-medium">Total:</p>
      <p className="text-2xl font-medium">&pound; {total.toFixed(2)}</p>
    </div>
  );
}
