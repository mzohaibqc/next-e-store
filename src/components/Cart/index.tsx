'use client';

import { useCart } from '@/store';

export default function Cart() {
  const { total } = useCart();
  return (
    <div className="w-full lg:w-[300px] lg:h-[400px] p-4 rounded-md border bg-white flex justify-between">
      <p className="text-lg font-medium">Cart Total:</p>
      <p className="text-2xl font-medium">
        &pound;<span data-testid="cart-total">{total.toFixed(2)}</span>
      </p>
    </div>
  );
}
