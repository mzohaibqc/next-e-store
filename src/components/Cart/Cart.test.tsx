import { fireEvent, render, screen } from '@testing-library/react';
import { expect, describe, test, vi } from 'vitest';
import Cart from '@/components/Cart';
import * as store from '@/store';

const cartItem: store.CartItem = {
  id: 1,
  name: 'Product 1',
  price: 10,
  img: '/product-1.jpg',
  colour: 'Black',
  count: 10,
};

describe('Cart', () => {
  const useCartSpy = vi.spyOn(store, 'useCart');
  useCartSpy.mockReturnValue({
    total: 100,
    cart: [cartItem],
    addItem: vi.fn(),
    removeItem: vi.fn(),
    clearItem: vi.fn(),
  });
  test('Cart should show the total price', () => {
    const { container } = render(<Cart />);
    expect(screen.getByTestId('cart-count').textContent).toEqual(
      cartItem.count.toString()
    );
    fireEvent.click(screen.getByTestId('cart-button'));
    expect(screen.getByTestId('cart-total').textContent).toEqual(
      (cartItem.price * cartItem.count).toFixed(2)
    );
    expect(container).toMatchSnapshot();
  });
});
