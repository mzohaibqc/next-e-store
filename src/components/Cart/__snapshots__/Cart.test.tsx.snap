// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cart > Cart should show the total price 1`] = `
<div>
  <div
    class="relative"
    data-headlessui-state="open"
  >
    <div
      aria-controls="headlessui-popover-panel-:r2:"
      aria-expanded="true"
      data-headlessui-state="open"
      id="headlessui-popover-button-:r0:"
    >
      <button
        class="text-gray-900 dark:text-gray-200 hover:bg-gray-100 dark:hover:bg-gray-700 focus:outline-none focus:ring-4 focus:ring-gray-200 dark:focus:ring-gray-700 rounded-md text-sm p-1 inline-flex items-center justify-center w-8 h-8"
        data-testid="cart-button"
      >
        <svg
          class="w-6 h-6 dark:text-gray-200 text-gray-900"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 20 20"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1 1.75A.75.75 0 011.75 1h1.628a1.75 1.75 0 011.734 1.51L5.18 3a65.25 65.25 0 0113.36 1.412.75.75 0 01.58.875 48.645 48.645 0 01-1.618 6.2.75.75 0 01-.712.513H6a2.503 2.503 0 00-2.292 1.5H17.25a.75.75 0 010 1.5H2.76a.75.75 0 01-.748-.807 4.002 4.002 0 012.716-3.486L3.626 2.716a.25.25 0 00-.248-.216H1.75A.75.75 0 011 1.75zM6 17.5a1.5 1.5 0 11-3 0 1.5 1.5 0 013 0zM15.5 19a1.5 1.5 0 100-3 1.5 1.5 0 000 3z"
          />
        </svg>
      </button>
      <span
        class="text-xs bg-red-600 text-white rounded-full w-5 h-5 flex justify-center items-center absolute -top-2 left-6"
        data-testid="cart-count"
      >
        10
      </span>
    </div>
    <div
      class="absolute left-1/2 z-10 mt-3 sm:mt-2 w-[400px] max-w-xs sm:max-w-sm -translate-x-full ml-8 transform px-4 sm:px-0 transition ease-out duration-200 opacity-0 translate-y-1"
      data-headlessui-state="open"
      id="headlessui-popover-panel-:r2:"
      tabindex="-1"
    >
      <div
        class="w-full p-4 rounded-md border border-gray-100 dark:border-gray-800 bg-white dark:bg-gray-900 flex flex-col justify-between gap-2 shadow-[0_8px_30px_rgb(0,0,0,0.12)]"
      >
        <div
          class="flex flex-col gap-2"
        >
          <div
            class="flex flex-col"
            data-testid="product-1"
          >
            <div
              class="flex-1 flex rounded-md overflow-hidden bg-white dark:bg-gray-900 shadow-[rgba(17,_17,_26,_0.1)_0px_0px_16px] border border-gray-200 dark:border-gray-800"
            >
              <img
                alt="Product Image"
                class="rounded-l-md"
                data-nimg="1"
                data-testid="product-1-img"
                decoding="async"
                fetchpriority="high"
                height="80"
                src="/_next/image?url=%2Fproduct-1.jpg&w=128&q=75"
                srcset="/_next/image?url=%2Fproduct-1.jpg&w=64&q=75 1x, /_next/image?url=%2Fproduct-1.jpg&w=128&q=75 2x"
                style="color: transparent; object-fit: cover; object-position: 50% 20%;"
                width="50"
              />
              <div
                class="info flex-1 px-2 py-1 flex flex-col"
              >
                <h2
                  aria-label="Product 1"
                  class="text-sm font-semibold mb-1 line-clamp-1"
                  title="Product 1"
                >
                  Product 1
                </h2>
                <div
                  class="flex justify-between items-center gap-3 mb-1"
                >
                  <div
                    class="flex items-center gap-3"
                  >
                    <div
                      class="rounded-md bg-white shadow-[0_8px_30px_rgb(0,0,0,0.12)] dark:border dark:border-gray-700 w-4 h-4"
                      data-testid="product-1-color"
                      style="background-color: rgb(0, 0, 0);"
                    >
                      <span
                        class="sr-only color-value"
                      >
                        Black
                      </span>
                    </div>
                    <p
                      class="text-sm text-gray-500 font-medium font-mono"
                    >
                      £
                      <span
                        data-testid="product-1-price"
                      >
                        10.00
                      </span>
                    </p>
                  </div>
                  <button
                    class="px-1 text-sm text-gray-600 disabled:text-gray-400 disabled:cursor-not-allowed disabled:hover:no-underline hover:underline focus:underline underline-offset-4 decoration-2"
                    data-testid="product-1-clear"
                  >
                    <span>
                      Remove
                    </span>
                  </button>
                </div>
                <div
                  class="w-full border border-gray-200 dark:border-gray-800 px-2 rounded-md sm:max-w-xs flex justify-around items-center"
                >
                  <button
                    aria-label="Remove One Item"
                    class="text-gray-900 dark:text-gray-200 hover:bg-gray-100 dark:hover:bg-gray-700 focus:outline-none focus:ring-4 focus:ring-gray-200 dark:focus:ring-gray-700 rounded-md text-sm p-1 inline-flex items-center justify-center w-5 h-5"
                    data-testid="product-1-remove"
                  >
                    <svg
                      class="w-4 h-4"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 448 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M416 208H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h384c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                      />
                    </svg>
                  </button>
                  <span
                    class="text-md text-gray-500 px-5 min-w-16 rounded-md text-center"
                    data-testid="product-1-count"
                  >
                    10
                  </span>
                  <button
                    aria-label="Add One Item"
                    class="text-gray-900 dark:text-gray-200 hover:bg-gray-100 dark:hover:bg-gray-700 focus:outline-none focus:ring-4 focus:ring-gray-200 dark:focus:ring-gray-700 rounded-md text-sm p-1 inline-flex items-center justify-center w-5 h-5"
                    data-testid="product-1-add"
                  >
                    <svg
                      class="w-4 h-4"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 448 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                      />
                    </svg>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="flex justify-between pt-2 px-1 border-t mt-3 "
        >
          <p
            class="text-lg font-medium"
          >
            Cart Total:
          </p>
          <p
            class="text-2xl font-medium font-mono"
          >
            £
            <span
              data-testid="cart-total"
            >
              100.00
            </span>
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;
